File: logitrack_web/src/utils/claimEligibility.ts
import type { ClaimType } from "../api/publicTracking";

export const CLAIM_INELIGIBLE_MESSAGE =
  "Tu envío todavía no ha sido entregado. Vas a poder reclamar este tipo una vez entregado.";

// Mínimo shape requerido para evaluar elegibilidad. Tanto `PublicShipment`
// (utilizado por el formulario público) como el `Shipment` del chatbot
// cumplen este contrato, así que la regla vive en un solo lugar.
export interface ClaimEligibilityShipment {
  status: string;
  estimated_delivery_at?: string | null;
}

// canFileClaim devuelve true cuando el envío permite reclamos generales:
// fue entregado o ya pasó al menos 1 hora desde la fecha estimada de entrega.
// Mantiene el mismo criterio que `canFileClaim` en
// logitrack_core/internal/handler/chatbot_handler.go.
export function canFileClaim(shipment: ClaimEligibilityShipment): boolean {
  if (shipment.status === "delivered") return true;
  const eta = shipment.estimated_delivery_at;
  if (!eta) return false;
  const deadline = new Date(eta).getTime() + 60 * 60 * 1000;
  return Date.now() > deadline;
}

// canFileClaimOfType extiende la regla general: maltrato es siempre reclamable
// desde que el envío existe (no depende del estado ni del SLA), porque puede
// ocurrir en cualquier interacción. El resto de los tipos respeta canFileClaim.
export function canFileClaimOfType(
  shipment: ClaimEligibilityShipment,
  claimType: ClaimType,
): boolean {
  if (claimType === "bad_treatment") return true;
  return canFileClaim(shipment);
}

File: logitrack_core/internal/service/claim_eligibility.go
package service

import (
	"time"

	"github.com/logitrack/core/internal/clock"
	"github.com/logitrack/core/internal/model"
)

// ClaimIneligibleMessage es el texto informativo que se devuelve al cliente
// cuando intenta abrir un reclamo de un tipo que el envío todavía no permite.
// Debe coincidir con CLAIM_INELIGIBLE_MESSAGE en
// logitrack_web/src/utils/claimEligibility.ts.
const ClaimIneligibleMessage = "Tu envío todavía no ha sido entregado. Vas a poder reclamar este tipo una vez entregado."

// CanFileClaim devuelve true cuando el envío permite reclamos "generales"
// (cualquier tipo distinto de maltrato): fue entregado o ya pasó al menos
// 1 hora desde la fecha estimada de entrega.
func CanFileClaim(shipment model.Shipment) bool {
	if shipment.Status == model.StatusDelivered {
		return true
	}
	if shipment.EstimatedDeliveryAt != nil {
		deadline := shipment.EstimatedDeliveryAt.Add(1 * time.Hour)
		if clock.Now().After(deadline) {
			return true
		}
	}
	return false
}

// CanFileClaimOfType es la única fuente de verdad de la regla de elegibilidad
// por tipo de reclamo. Maltrato (bad_treatment) puede iniciarse desde que el
// envío existe, sin importar estado ni SLA, porque puede ocurrir en cualquier
// interacción con el personal. El resto de los tipos respeta CanFileClaim.
// Debe mantenerse en sync con canFileClaimOfType del frontend.
func CanFileClaimOfType(shipment model.Shipment, claimType model.ClaimType) bool {
	if claimType == model.ClaimTypeBadTreatment {
		return true
	}
	return CanFileClaim(shipment)
}
